[aoti] Remove dir after packaging

angelayi opened the pull request:
Update AOTI to return a list of files that it generates when `aot_inductor.package=True`. Then we will only package the files that are in that list. 

This should fix the [caching issue](https://fb.workplace.com/groups/1028545332188949/permalink/1081702043539944/) and hopefully https://github.com/pytorch/pytorch/issues/140053.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

desertfire reviewed (changes requested):
This doesn't solve the problem when different runs can generate different cubin files. We can end up with including unnecessary cubin files.

I think a better way to solve this is in codecache.py. Using an unique subdirectory to store .so and other relevant files and package afterwards. This way, all the previous auto-tuning results are still kept and we will not package unnecessary files.

desertfire commented:
Also your test script should still be added as a unit test. It should work with some tweaks.

desertfire commented:
> This doesn't solve the problem when different runs can generate different cubin files. We can end up with including unnecessary cubin files.
> 
> I think a better way to solve this is in codecache.py. Using an unique subdirectory to store .so and other relevant files and package afterwards. This way, all the previous auto-tuning results are still kept and we will not package unnecessary files.

I see you are actually deleting the whole directory afterwards. This doesn't solve the keep caching request that Henry raised.

henrylhtsang commented:
This can unblock (previously will run into errors). But would like to see a way to cache things so iteration speed can be better. 

For torch.compile, subsequent runs takes 1/10 of the time to compile due to local cache. It would be nice if AOTI can have similar features.

desertfire commented on torch/_inductor/compile_fx.py:
```diff
@@ -1236,14 +1235,28 @@ def compile_fx_aot(
+            "If you would like to package the AOTInductor generated files "
+            "into a pt2, please call `torch._inductor.aoti_compile_and_package`."
+        )
+    elif config.aot_inductor.output_path:
```
Why not using an `or` condition with the previous `elif`?

desertfire commented on torch/_inductor/package/package.py:
```diff
@@ -165,35 +167,59 @@ def package_aoti(archive_file: str, aoti_files: Union[str, Dict[str, str]]) -> s
+                    filename,
+                )
+                archive_writer.write_file(
+                    f"{AOTINDUCTOR_DIR}{model_name}/{filename}",
```
Let's use utility functions from `os.path` for path concatenation and split (Line 210).

desertfire reviewed (commented):
LGTM overall.

The CI "out of disk" failure is real. I think it's related to the fact that when we call `aoti_load_package`, we unzip the files to the top level `/tmp` directory, which will not removed after running each benchmark. The large weight files gradually eat the disk and eventually triggers the error.

malfet commented on torch/_inductor/__init__.py:
```diff
@@ -96,9 +96,12 @@ def aoti_compile_and_package(
 
     if package_path is None:
-        package_path = aoti_files + ".pt2"
+        package_path = f"{so_path[:-3]}.pt2"
```
Either use Pathlib tools or rsplit, which is much safer that hardcoding the suffix size (unless you want to assert beforehand that `so_path.ends_with(".so")`)
```suggestion
        package_path = f"{so_path.rsplit('.', 1)[0]}.pt2"
```

malfet commented on torch/_inductor/__init__.py:
```diff
@@ -143,7 +146,8 @@ def aot_compile(
 
     Returns:
-        Path to the generated shared library
+        Path to the generated shared library, or a list of files generated by
```
This feels like a complicated API, why not always return a list, albeit for .so it will be a single element list

angelayi replied:
I don't want to break any existing callsites 😅 but yes! I can fix the rest of the callsites to return a list in a followup.

malfet commented on torch/_inductor/package/package.py:
```diff
@@ -165,35 +167,56 @@ def package_aoti(archive_file: str, aoti_files: Union[str, Dict[str, str]]) -> s
+                if file == "":
+                    continue
+
+                if file.endswith(".so"):
```
Perhaps a discussion for different PR, but should it really be `.so` on all platforms? Wouldn't it be more reasonable to check for `sysconfig.get_config_var('EXT_SUFFIX')`?

malfet commented on torch/_inductor/codegen/cpp_wrapper_gpu.py:
```diff
@@ -288,22 +304,21 @@ def generate_load_kernel_once(
+        deferred_gpu_kernel_line = DeferredGpuKernelLine(
+            kernel_name,
+            (
+                """    """
+                + kernel_var_name
+                + """ = loadKernel("%s", "%s", %s, this->cubin_dir_);"""
+                if V.graph.aot_mode
+                else """    """ + kernel_var_name + """ = loadKernel("%s", "%s", %s);"""
```
I know you just moving code around, but please consider at least replacing `""" """` with `" "` (and may be use f-strings)

agunapal commented:
@angelayi What are the contents of the packaged artifact? Do we store the `.cubin` files too?  Is this for caching of the compiled results?

henrylhtsang commented on torch/_inductor/graph.py:
```diff
@@ -2026,9 +2026,15 @@ def compile_to_fn(self) -> Any:
                     serialized_extern_kernel_nodes,
                 )
 
+            additional_files = self.wrapper_code.additional_files
```
noob question: when we add strings to generated_files, we will also add the strings to self.wrapper_code.additional_files. Is this a concern?

desertfire commented:
> @angelayi What are the contents of the packaged artifact? Do we store the `.cubin` files too? Is this for caching of the compiled results?

It does contain .cubin files, and a few other metadata files. It's a more systematic way to distribute the AoT compiled artifact. Previously people may blindly copy .so file without realizing .cubin files are also needed.

angelayi commented:
@pytorchbot merge -f "failures unrelated, i think"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but the lint failure is legit' -c ignoredsignal

kit1980 commented:
Thanks for reverting, I was about to revert for internal failures (see D65607978)

angelayi commented:
@pytorchbot merge

angelayi commented:
@pytorchbot revert -m 'sorry I realized I need to land from internal' -c codev

angelayi commented:
@pytorchbot revert -m 'sorry I realized I need to land from internal' -c ghfirst

angelayi committed:
[aoti] Remove dir after packaging (#140022)

Summary:
Update AOTI to return a list of files that it generates when `aot_inductor.package=True`. Then we will only package the files that are in that list. 

This should fix the [caching issue](https://fb.workplace.com/groups/1028545332188949/permalink/1081702043539944/) and hopefully https://github.com/pytorch/pytorch/issues/140053.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov


Reviewed By: pianpwk

Differential Revision: D65862850

Pulled By: angelayi
